torch/onnx/utils.py

```diff
@@ -137,23 +132,45 @@ def _export(model, args, f, export_params=True, verbose=False, training=False,
         raise RuntimeError("state_dict changed after running the tracer; "
                            "something weird is happening in your model!")
 
-    trace.set_graph(_optimize_graph(trace.graph(), aten))
+    return trace.graph(), torch_out
+
+
+# NOTE: the output `torch_out` will contain the output tensors resulting from
+# the trace of a Module. In the case that a torch.nn.ScriptModule is passed in,
+# this output will be None, since we are not doing any tracing but rather
+# directly extracting the graph.
+def _export(model, args, f, export_params=True, verbose=False, training=False,
+            input_names=None, output_names=None, aten=False, export_type=ExportTypes.PROTOBUF_FILE):
+    # Special case for common case of passing a single Variable
+    if isinstance(args, torch.autograd.Variable):
+        args = (args, )
+
+    if isinstance(model, torch.jit.ScriptModule):
+        torch_out = None
+        try:
+            graph = model.__getattr__('forward').graph()
+            graph.propagate_shapes(args, False)
+        except AttributeError:
+            # TODO: just trace it
+            raise RuntimeError('\'forward\' method must be a script method')
```

Comment:
Can you fix that? It's actually really simple now. Just assing `graph = None` before this case, and change the `else` to `if graph is None`

Reply:
It's not as simple as that right now since ScriptModules don't support things like `set_training` and `slow_forward`. I didn't want to pull all that into this PR

cc @zdevito

Reply:
We don't need `slow_forward`, because we can just inline the graph, right?

Also, good point about the `.training` attribute. We absolutely need to support it at some point.

Reply:
I think it is fine to open issues for resolving these other cases. But you _should_ be able to trace a non-script function once the issues like the `.training` attribute are fixed.